### PR TITLE
JAPI-178 DFSClient: Null pointer exception when reading some key files

### PR DESCRIPTION
**Pull request by jpmcmu**
- Added code to keep track of flags on Record fields
- Modified RecordDefinitionTranslator to setAdditionalFlags and use additional flags when generating JSON record defs
- Improved exception messages to make actual errors more evident

Signed-off-by: James McMullan <James.McMullan@lexisnexis.com>

## Type of change:
- [X] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).

## Checklist:
- [X] I have created a corresponding JIRA ticket for this submission
- [X] My code follows the code style of this project.
  - [ ] I have applied the Eclipse code-format template provided.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [X] I have read the HPCC Systems CONTRIBUTORS document (https://github.com/hpcc-systems/HPCC-Platform/wiki/Guide-for-contributors).
- [X] The change has been fully tested:
  - [ ] I have performed unit tests to cover my changes.
  - [ ] I have performed system test and covered possible regressions and side effects.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [X] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.

### CODE REVIEW

**Comment by jpmcmu** [comment]
@rpastrana please review

File: commons-hpcc/src/main/java/org/hpccsystems/commons/ecl/FieldDef.java [hunk 1]
```diff
@@ -63,16 +65,47 @@ public FieldDef(FieldDef rhs)
...
+     * @param isUnsigned
+     *            Only applies to integers and decimal fields. Ignored otherwise
+     * @param HpccSrcType
+     *            Field encoding type. Primairly applies to strings. Non-strings should use LITTLE_ENDIAN.
+     * @param childDefs
+     *            Child field defs. Only used for Records, Sets & Datasets. Sets & Dataset should have a single child. Null otherwise.
      */
     public FieldDef(String fieldName, FieldType fieldType, String typeName, long len, boolean isFixedLength,
-            boolean isUnsigned, HpccSrcType styp, FieldDef[] defs)
+            boolean isUnsigned, HpccSrcType styp, FieldDef[] childDefs)
```
Comment (rpastrana) [inline commons-hpcc/src/main/java/org/hpccsystems/commons/ecl/FieldDef.java]: very minor, but we should come up w a better name for styp

File: commons-hpcc/src/main/java/org/hpccsystems/commons/ecl/FieldDef.java [hunk 2]
```diff
@@ -35,6 +35,7 @@
     private long              len                = 0;
     private boolean           fixedLength        = false;
     private boolean           isUnsigned         = false;
+    private int               additionalFlags    = 0;
```
Comment (rpastrana) [inline commons-hpcc/src/main/java/org/hpccsystems/commons/ecl/FieldDef.java]: is there a known set of valid additional flags?

File: commons-hpcc/src/main/java/org/hpccsystems/commons/ecl/RecordDefinitionTranslator.java [hunk 3]
```diff
@@ -137,6 +141,12 @@ private static boolean isUnsigned(int typeID)
         return (typeID & FLAG_UNSIGNED) != 0;
     }
 
+    private static int getAdditionalFlags(int flags)
```
Comment (rpastrana) [inline commons-hpcc/src/main/java/org/hpccsystems/commons/ecl/RecordDefinitionTranslator.java]: Is this method for callers interested only in the 0x8000 bit, or are would they also be interested in the higher bits

File: commons-hpcc/src/main/java/org/hpccsystems/commons/ecl/FieldDef.java (hunk 2 above)
Comment (jpmcmu) [reply commons-hpcc/src/main/java/org/hpccsystems/commons/ecl/FieldDef.java]: At the moment there is only one flag that we need to give to dafilesrv, but we do not use ourselves. However, the platform code is structured in such a way that there could potentially be more flags that fall into this "track only" category. All of these flags would be 0x8000 or above.

**Comment by jpmcmu** [comment]
@rpastrana code review changes pushed up. please review

**Pull request merged**